src/libraries/Native/Unix/System.Native/pal_random.c

```diff
@@ -20,91 +21,230 @@ Generate random bytes. The generated bytes are not cryptographically strong.
+{
+    static bool sSyscallNotSupported;
+
+    if (sSyscallNotSupported)
     {
-        if (rand_des == -1)
-        {
-            int fd;
+        return false;
+    }
+
+    for (;;)
+    {
+        long bytesFilled = getrandom(buffer, bufferLength, 1 /* GRND_NONBLOCK */);
 
-            do
+        if (bytesFilled == -1)
+        {
+            switch (errno)
             {
+                case EINTR:
+                    continue;
+                case ENOSYS:
+                    sSyscallNotSupported = true;
+                    // Fallthrough
+                default:
+                    return false;
+            }
+        }
+
+        if (bytesFilled == (long)bufferLength)
+        {
+            return true;
+        }
+
+        if (bytesFilled < (long)bufferLength)
+        {
+            bufferLength -= (size_t)bytesFilled;
+            buffer += bytesFilled;
+        }
+    }
+}
+
+#else // !HAVE_ARC4RANDOM_BUF && !HAVE_SYS_GETRANDOM
+
+#define MAY_NEED_XOR_WITH_LRAND 1
+static bool PlatformRandomFillBuffer(uint8_t *buffer, size_t bufferLength)
+{
+    (void)buffer;
+    (void)bufferLength;
+
+    return false;
+}
+
+#endif
+
+static bool PlatformRandomFillBufferUrandomFallback(uint8_t *buffer, size_t bufferLength)
+{
+    static int sUrandomFd = -1;
+    static bool sDevUrandomIsUnusable;
+
+    if (sDevUrandomIsUnusable)
+    {
+        return false;
+    }
+
+    if (sUrandomFd < 0)
+    {
+        int fd;
+
+        do
+        {
 #if HAVE_O_CLOEXEC
-                fd = open("/dev/urandom", O_RDONLY, O_CLOEXEC);
+            fd = open("/dev/urandom", O_RDONLY | O_CLOEXEC);
 #else
-                fd = open("/dev/urandom", O_RDONLY);
-                fcntl(fd, F_SETFD, FD_CLOEXEC);
+            fd = open("/dev/urandom", O_RDONLY);
+            fcntl(fd, F_SETFD, FD_CLOEXEC);
 #endif
-            }
-            while ((fd == -1) && (errno == EINTR));
+        }
+        while ((fd < 0) && (errno == EINTR));
 
-            if (fd != -1)
+        if (fd >= 0)
+        {
+            int expected = -1;
+
+            if (!__atomic_compare_exchange_n(&sUrandomFd, &expected, fd, false, __ATOMIC_SEQ_CST, __ATOMIC_SEQ_CST))
             {
-                int expected = -1;
-                if (!__atomic_compare_exchange_n(&rand_des, &expected, fd, false, __ATOMIC_SEQ_CST, __ATOMIC_SEQ_CST))
-                {
-                    // Another thread has already set the rand_des
-                    close(fd);
-                }
+                // Another thread set sUrandomFd
+                close(fd);
             }
-            else if (errno == ENOENT)
-            {                
-                sMissingDevURandom = true;
+        }
+        else
+        {
+            switch (errno) {
+            case EACCES:
+            case EPERM:
+            case ENOENT:
+                sDevUrandomIsUnusable = true;
+                // Fallthrough
+            default:
+                return false;
```

Comment:
What errors might happen where we don't want to mark sDevUrandomIsUnusable = true?

Reply:
Transient errors such as `EINTR`, mostly.  `open()` can return way too many errors, and most of them are unlikely to happen in something such as when trying to open `/dev/urandom`; it's cheap to keep trying to open.  If anything, it's not worse than what it is already doing.